### ISSUE TITLE
RFE: add sd_bus_slot cleanup callback

### ISSUE THREAD

**Issue by daurnimator** [systemd/systemd#415]
Continuing from https://github.com/systemd/systemd/issues/327

I proposed:

> Associating some sort of cleanup callback with a slot would be great.

@poettering said

> i'd be willing to take a patch that adds a destroy callback logic to slots, I'd be happy to take it. But I think that should be discussed in a separate issue

**Comment by dvdhrm** [comment]
Please describe this in full detail. It's cumbersome to jump between threads to gather information. I appreciate the links, but I really wouldn't mind duplicating a description of your issue.

Anyhow, sd_bus_slot cleanup callbacks sound easy to implement. But I need further information:
- when should it be called
- what context should it have
- _why_ is it exactly needed
- _who_ is going to use it

The more elaborate you are, the easier it is to figure out whether this really makes sense.

**Comment by daurnimator** [comment]
> Please describe this in full detail. It's cumbersome to jump between threads to gather information. I appreciate the links, but I really wouldn't mind duplicating a description of your issue.

That's pretty much it:
To add some callback that is called when an `sd_bus_slot` is destroyed. 

> when should it be called

inside of `bus_slot_disconnect`

> what context should it have

just the slot is fine I think, once you have it you can:
- grab the bus (`sd_bus_slot_get_bus`) out of it
- get the userdata (`sd_bus_slot_get_userdata`)

> why is it exactly needed

To clean up (e.g. `free`) the userdata stored in the object
Invalidate any weak references to the slot

> who is going to use it

anyone using sd-bus and tieing resources to a slot.

**Comment by dvdhrm** [comment]
I don't see why that's needed, sorry. I mean, why not just destroy your userdata at the time you unref your slot-reference? Sure, the object might still be alive as someone else might have a ref. But then again, just don't hand out references to your slot.

The same issue exists with sd_event_source: just because it is ref-counted does not mean that you need a destroy hook. In fact, most systemd code relies on *_unref() to actually destroy the object. And this is fine as long as you don't hand out references to external code (destruction might be delayed until the own callback finishes, in case of sd-slot).

Sure, it is questionable whether ref-counts make sense for those objects. But it's how it is. If you really cannot be sure your *_unref() call destroys the object, reset userdata to NULL and check for it in your callback.

I'm still wondering what a real use-case for a destroy-hook would be..
Not convinced, sorry.

**Comment by daurnimator** [comment]
> I don't see why that's needed, sorry. I mean, why not just destroy your userdata at the time you unref your slot-reference? Sure, the object might still be alive as someone else might have a ref. 

So you're suggesting that the proper way to `unref` is to `unref`, then `sd_bus_slot_set_userdata`?.

There is still no way to remove my callback: my message handler is in a dynamically loaded object that I'd like to dlclose once it's not needed any more...

> But then again, just don't hand out references to your slot.

AFAIK, there's no way to stop a library or thread from using `sd_bus_get_current_slot` on a bus. this particulary matters for the default bus.

> The same issue exists with sd_event_source

I haven't even looked into `sd_event_*` yet... infact I didn't know they had been exposed.

> And this is fine as long as you don't hand out references to external code (destruction might be delayed until the own callback finishes, in case of sd-slot).

An issue is that it's possible for external code to get a ref without handing it out.
But also that the lifecycle is not documented. e.g. if I close the bus a slot belongs to, will the slot stay valid?

> But it's how it is

We can change it.....

> If you really cannot be sure your *_unref() call destroys the object, reset userdata to NULL and check for it in your callback.

Please also add something to NULL out the callback handler if this is the case.

**Comment by utezduyar** [comment]
I don't think you need to mess with sd_bus_slot_set_userdata. All the slots create a reference to the user data.

Like @dvdhrm said, shouldn't following be enough to clean up your stuff?

```
/* your server runs with a non-floating slot */
sd_bus_add_object_vtable(bus, slot, ..., userdata)
/* you decided to stop the service */
sd_bus_slot_unref(slot) /* We destroy the slot here, so no more ref to your user data */
sd_bus_unref (bus)
free (userdata)
```

**Comment by dvdhrm** [comment]
I agree with @utezduyar.

Sure, there _might_ be other code that owns a reference to your slot, but why would you care? I mean, you as the slot owner have full control over the slot. Don't pass references to other code, if you don't want it. Hence, you can be sure the slot is unused after you drop your reference.

In fact, I'd even say you're doing something horribly wrong if you call sd_bus_slot_ref() or sd_event_source_ref(). I cannot see a real usecase outside of libsystemd.

**Comment by daurnimator** [comment]
> Sure, there might be other code that owns a reference to your slot, but why would you care?

because then my callback function will end up segfaulting. or libsystemd will try and call my callback function which no longer exists and segfault. Or worse: no segfault, something else has been loaded in it's place and gets run instead.

e.g.
1. I grab a default bus
2. I call `sd_bus_add_filter` and get a slot.
3. I process some matches.
4. I unref my slot.
5. I `dlclose` the library my callback was in; the callback no longer exists
6. A dbus message comes in

Now imagine between steps 3 and 4, some other piece of the process refs the slot.
This means that when the message arrives in 6, you get a segfault at best, at worst, jump to a random piece of executable memory.

>  I mean, you as the slot owner have full control over the slot. Don't pass references to other code, if you don't want it. Hence, you can be sure the slot is unused after you drop your reference.

As far as I know, there's no way to be assured that another piece of code won't ref my slot; maybe it's a debugger? maybe someone just feels like calling `sd_bus_get_current_slot` and `ref`ing it?

But considering they can, I would like a way to be notified when the slot is destroyed. That way I can ensure that e.g.
- `dlclose` is not called until destruction
- my userdata is not freed

**Comment by utezduyar** [comment]
I don't get how callback will help you in this case though? If slot has many references it will never be cleaned up anyways. So, I don't get how that callback will help you. Also pay attention that library calls `bus_slot_disconnect` for floating slots (slots that are not referenced by the application).

As far as I can see `sd_bus_get_current_slot` is useful for the reply callbacks. Once you are in the callback, you can figure out which slot was registered to make the call. I have double checked that library is setting the `bus->current_slot` to NULL after the callback [1]. So, I don't think your concern of people can start calling  `sd_bus_get_current_slot` to get a reference to your slot is valid. 

[1] @dvdhrm I can see that _only_ `process_filter` is setting the current slot to `bus->current_slot = sd_bus_slot_unref(slot);` after the callback. Is this really correct? Shouldn't it be setting it to NULL and unrefing the slot in separate execution?

**Comment by daurnimator** [comment]
I'll need a few days to play around and come up with some examples

---

> [1] @dvdhrm I can see that only process_filter is setting the current slot to bus->current_slot = sd_bus_slot_unref(slot); after the callback. Is this really correct? Shouldn't it be setting it to NULL and unrefing the slot in separate execution?

`sd_bus_slot_unref` returns `NULL` in all cases, so I don't think that's an issue

**Comment by poettering** [comment]
I'd probably still at the destroy callbacks, and do that for all our objects, also sd_event_source ones... This is specifically useful for "floating" objects, where you don't keep the slot object around, but where the slot object is implicitly destroyed when the owning object goes away.

That said, we currently implement floating objects by taking NULL object pointers... To set a destroy callback on them you'd need them however, so this usecase is kinda contradicting our existing API... Maybe we should have cals to turn a normal slot or event source object into a floating one...

**Comment by dvdhrm** [comment]
@poettering: I still cannot see a real use-case. I cannot even see a use-case why slot-objects or event-sources are ref-counted (except for completeness). Unless there's no user of `sd_bus_slot_ref` nor `sd_event_source_ref`, I don't see the point.

(The exisiting users of `sd_bus_slot_ref` just make sure slots survive a callback. We could easily deploy the same workaround as we do for event-sources.)

(The only user of `sd_event_source_ref` (remote-journal) is actually a no-op and can just be removed.)

**Comment by poettering** [comment]
OK, so let's not do this for now. If anyone comes up with a strong usecase, we can reconsider this, but I have not seen a strong usecase so far that wasn't better solved otherwise.

**Comment by pszewczyk** [comment]
@poettering: I would really like to reopen this issue. I stumbled upon a case, which is hard to handle without this kind of callback.

I have a module in my code which handles some events and by result is doing `sd_bus_add_match` with different user data every time. I allocate it before `sd_bus_add_match` and free in the registered callback.
But now freeing all this allocated data when the bus is closing requires keeping track of every slot I created. But why should I do that if sd-bus is doing it already internally?

And it doesn't even concern refcounts, I also see no real reason for them here. But when I close and free my bus I would expect that all resources related to it will be cleaned up. If I consider userdata passed to the callback this kind of resource (and I do), the bus cleanup mechanism is not complete.
